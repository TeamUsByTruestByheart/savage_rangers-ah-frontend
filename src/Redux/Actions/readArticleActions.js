import types from './index';
import axios from '../../configs/axios';

const { READ_ARTICLE, GET_ARTICLE_TAGS } = types;

export default {
  readArticle: slug => (dispatch) => {
    axios
      .get(`https://authors-heaven.herokuapp.com/api/articles/${slug}`)
      .then((res) => {
        dispatch({
          type: READ_ARTICLE,
          payload: res.data.article
        });
      });
  },
  getTags: slug => (dispatch) => {
    axios
      .get(`https://authors-heaven.herokuapp.com/api/articles/${slug}/tags`)
      .then((res) => {
        dispatch({
          type: GET_ARTICLE_TAGS,
          payload: res.data
        });
      });
  }
};
