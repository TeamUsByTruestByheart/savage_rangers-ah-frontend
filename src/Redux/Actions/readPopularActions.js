import types from './index';
import axios from '../../configs/axios';

const { GET_POPULAR_ARTICLES, CATCH_ERROR } = types;

export default {
  readPopularArticle: () => (dispatch) => {
    axios
      .get('/api/articles/most/popular')
      .then((res) => {
        dispatch({
          type: GET_POPULAR_ARTICLES,
          payload: res.data
        });
      }).catch((err) => {
        const error = err.response.data.errors;
        dispatch({ type: CATCH_ERROR, payload: error });
      });
  }
};
