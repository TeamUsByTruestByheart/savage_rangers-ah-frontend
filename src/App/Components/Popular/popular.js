import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import truncate from 'truncate';
import { Link } from 'react-router-dom';
import ReactImageFallback from 'react-image-fallback';
import Actions from '../../../Redux/Actions/readPopularActions';


const { readPopularArticle } = Actions;
export class Popular extends Component {
  constructor(props) {
    super(props);
    this.state = { articles: [{ title: '', User: {}, Category: {} }] };
  }

  componentWillMount() {
    const { readPopularArticle: readPopular } = this.props;
    readPopular();
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ articles: nextProps.popularArticle.data });
  }

  render() {
    const { articles } = this.state;
    return (
      <div className="container article__popular">

        <p id="popular-title">Popular</p>
        <hr />
        {
          articles.map(item => (
            <Link to={`/articles/${item.slug}`} key={item.title} style={{ color: '#000', textDecoration: 'none' }}>
              <div className="article__popular-content">
                <ReactImageFallback className="content-image" src={item.coverImage} fallbackImage="https://ielektro.es/wp-content/uploads/2017/04/ventajas-comprar-LED.jpg" alt="content" title="popular Image" />
                <div className="holder">
                  <p className="holder__content-title">
                    {truncate(item.title, 31)}
                    ...
                  </p>
                  <div className="holder__vl" />
                  <p className="holder__category">{item.Category.name}</p>
                  <div className="holder__writer">
                    <ReactImageFallback className="holder__writer-img" src={item.User.profileImage} fallbackImage="https://ielektro.es/wp-content/uploads/2017/04/ventajas-comprar-LED.jpg" alt="content" title="Writer Image" />
                    <p className="writer__name">{`${item.User.firstName} ${item.User.lastName}`}</p>
                    <p className="writer__date">{moment(item.createdAt).format('Do MMM YYYY')}</p>
                  </div>

                </div>

              </div>
            </Link>

          ))
        }
      </div>
    );
  }
}

export const mapStateToProps = state => ({ popularArticle: state.populars.Articles });


export default connect(mapStateToProps, { readPopularArticle })(Popular);
