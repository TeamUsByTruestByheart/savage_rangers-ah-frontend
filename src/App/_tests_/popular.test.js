import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import mockAxios from 'axios';
import promiseMiddleware from 'redux-promise-middleware';
import { shallow } from '../../enzyme';
import { Popular, mapStateToProps } from '../Components/Popular/popular';
import populaArticleReducer from '../../Redux/Reducers/popularArticleReducer';
import readArticleActions from '../../Redux/Actions/readArticleActions';
import readPopularActions from '../../Redux/Actions/readPopularActions';

const { readArticle, getTags } = readArticleActions;
const { readPopularArticle } = readPopularActions;

const middlewares = [thunk, promiseMiddleware];
const mockStore = configureStore(middlewares);

const initialState = { Article: {} };
const store = mockStore(initialState);

const payload = {
  populars: {
    Popular: {},
    Articles: {
      data: [{
        id: 6,
        title: 'What you need to know to become a great software engineer in 2020',
        description: '10 Tips to become a great software engineer by 2020.',
        body: 'I get quite a few emails',
        slug: 'what-you-need-to-know-to-become-a-great-software-engineer-in-2020-nfovo9fghng',
        readTime: 6,
        coverImage: 'https://res.cloudinary.com/al-tech/image/upload/v1565787533/rvkmbae3pcqhgysxoqnk.jpg',
        author: 42,
        category: 4,
        isBlocked: false,
        status: 'published',
        createdAt: '2019-08-14T12:58:54.185Z',
        updatedAt: '2019-08-14T13:00:34.847Z',
        deletedAt: null,
        User: {
          id: 42, username: 'MCFrank16', profileImage: 'https://res.cloudinary.com/al-tech/image/upload/v1565805494/nenxg6yetjfgbk96iq8l.jpg', firstName: 'Frank', lastName: 'Mutabazi'
        },
        Category: { name: 'TECHNOLOGY' }
      }]
    }
  }
};

mapStateToProps({ populars: {} });
const props = { readPopularArticle: jest.fn() };
const popular = shallow(<Popular {...props} />);
describe('Popular component', () => {
  it('renders the Popular component', () => {
    expect(popular).toBeDefined();
    expect(popular.find('div').exists()).toBe(true);
    expect(popular).toMatchSnapshot();
  });
  it('receive the props', () => {
    popular.setState({ articles: [] });
    popular.instance().componentWillReceiveProps({ popularArticle: { data: [] } });
    expect(popular.instance().props).toEqual(props);
    expect(popular.instance().state).toEqual({ articles: [] });
  });
});

it('should get articles with tags', () => {
  store.dispatch(readArticle());
  store.dispatch(getTags());
  expect(store.getActions()).toEqual([]);
});

it('should get popular articles', () => {
  mockAxios.get = jest.fn(() => Promise.resolve());
  store.dispatch(readPopularArticle());
  expect(store.getActions().length).toEqual(2);
});

test('reducers', () => {
  const state = populaArticleReducer(initialState,
    {
      type: 'GET_POPULAR_ARTICLES',
      payload
    });

  expect(state).toEqual({
    Article: {},
    Articles: { ...payload }
  });
});

describe('testing the get popular article action', () => {
  it('should get all popular articles when fetching them is properly executed', () => {
    const dataTest = { body: 'this is awesome' };
    mockAxios.get = jest.fn(() => Promise.reject({ response: { data: { errors: {} } } }));
    store.dispatch(readPopularArticle(dataTest));
    expect(store.getActions().length).toEqual(2);
  });
});
