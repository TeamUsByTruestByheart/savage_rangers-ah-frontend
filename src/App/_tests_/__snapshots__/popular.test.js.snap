// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popular component renders the Popular component 1`] = `ShallowWrapper {}`;
